transformers/hf_api.py

```diff
@@ -148,6 +152,27 @@ def list_objs(self, token):
 
 
 
+class ProgressFileReader:
```

Comment:
Definitely non-standard, but definitely clever :)

I don't think many people would understand this behavior, as you have to know values are passed by reference in Python, and be familiar with `io.BufferedReader` 's API. I would document the `ProgressFileReader` class to explain why we are doing this and why it works.

A more pythonic API would be to promote `ProgressFileReader` to a [context manager](https://docs.python.org/3/library/contextlib.html)  and open the file in this context manager so the uploading code would look like:

```python
with stream_with_progressbar(filepath, "rb") as f:
       r = requests.put(urls.write, data=f, headers={
              "content-type": urls.type,
       })
       r.raise_for_status()
```

Also has the merit of being self-documenting. The closing logic is handled inside the context manager.

Reply:
I could do a context manager, but I can't encapsulate `f` completely like in your code sample, (passing `f` to `requests.put`) as I don't really know which methods of the io.BufferedReader `requests` (and its underlying libraries) is going to call.

The cleanest way would be to subclass `io.BufferedReader` and instantiate an instance of the child class with the parent instance, but I don't know how to do that in Python.